### PR TITLE
Phase 6: SGDR Warm Restarts — Cosine Annealing with Restarts

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

SGDR (Loshchilov & Hutter, ICLR 2017) uses cosine annealing with periodic warm restarts. Each restart drives the LR back up, allowing the model to escape local minima and explore new loss basins. The `scheduler_type="warm_restarts"` flag is already implemented in train.py but has not been tested with our current baseline.

**Why this might help:** p_tan=29.1 is significantly worse than p_in=12.2, suggesting the model overfits to the single-foil geometry. Warm restarts encourage more exploration of the loss landscape, potentially finding basins that generalize better to the tandem-foil OOD split.

**Key timing with EMA:** With T_0=40 and T_mult=2:
- Warmup: epochs 0-9 (10 epochs, hardcoded start_factor=0.1)
- Cycle 1: epochs 10-49 (40 epochs)
- Cycle 2: epochs 50-129 (80 epochs)
- Cycle 3 starts at epoch 130
- EMA starts at epoch 140 — 10 epochs into the final cycle's descent

This is favorable: restarts encourage exploration, EMA stabilizes the final descent into the best basin.

### No code changes needed — just flag changes

The warm restarts scheduler is fully implemented. Use these flags:
```
--scheduler_type warm_restarts --cosine_T_0 <T0> --cosine_T_mult 2
```

### Experiment matrix (8 GPUs)

| GPU | Config | Seed | Notes |
|-----|--------|------|-------|
| 0 | Baseline (sequential cosine) | 42 | Control |
| 1 | Baseline (sequential cosine) | 43 | Control |
| 2 | `--scheduler_type warm_restarts --cosine_T_0 40 --cosine_T_mult 2` | 42 | Main test |
| 3 | `--scheduler_type warm_restarts --cosine_T_0 40 --cosine_T_mult 2` | 43 | Main test |
| 4 | `--scheduler_type warm_restarts --cosine_T_0 20 --cosine_T_mult 2` | 42 | Faster restarts |
| 5 | `--scheduler_type warm_restarts --cosine_T_0 20 --cosine_T_mult 2` | 43 | Faster restarts |
| 6 | `--scheduler_type warm_restarts --cosine_T_0 60 --cosine_T_mult 2` | 42 | Slower restarts |
| 7 | `--scheduler_type warm_restarts --cosine_T_0 60 --cosine_T_mult 2` | 43 | Slower restarts |

Use `--wandb_group phase6/sgdr-warm-restarts` for all runs.

**SGDR command template:**
```bash
python train.py --agent askeladd --wandb_group phase6/sgdr-warm-restarts \
  --wandb_name "askeladd/sgdr-T0${T0}-s${seed}" \
  --scheduler_type warm_restarts --cosine_T_0 ${T0} --cosine_T_mult 2 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed ${seed}
```

Note: Remove `--cosine_T_max 160` — it's only used by the sequential scheduler and is ignored by warm_restarts.

**Baseline command:** Same as above but WITHOUT `--scheduler_type warm_restarts --cosine_T_0 ... --cosine_T_mult ...` (uses default sequential scheduler with `--cosine_T_max 160`).

### Important notes
- DO NOT modify MAX_EPOCHS (500) or MAX_TIMEOUT (180.0) in train.py
- The warm_restarts scheduler has its own 10-epoch warmup (start_factor=0.1, hardcoded) — different from the sequential scheduler's 20-epoch warmup
- `cosine_eta_min` defaults to 1e-5 and applies to warm_restarts too

## Baseline

Current best (8-seed ensemble, PR #2080):
- p_in: **12.2** | p_oodc: **6.7** | p_tan: **29.1** | p_re: **5.8**

Single-model baseline (8-seed mean):
- p_in: **13.03** | p_oodc: **7.83** | p_tan: **30.29** | p_re: **6.45**

Compare SGDR results against the SINGLE-MODEL baseline mean.

## Reporting

Post a comment with results in table format including:
- Individual run metrics (p_in, p_oodc, p_tan, p_re) for all 8 runs
- Mean ± std for each T_0 setting vs baseline
- W&B run IDs
- Which T_0 value worked best and why

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Individual Run Metrics

| GPU | Config | Seed | val/loss | p_in | p_oodc | p_tan | p_re | W&B Run |
|-----|--------|------|----------|------|--------|-------|------|---------|
| 0 | Baseline (sequential) | 42 | 0.3899 | 13.3 | 7.6 | 31.0 | 6.4 | dk8vk0o1 |
| 1 | Baseline (sequential) | 43 | 0.3883 | 12.7 | 8.1 | 30.4 | 6.6 | jz89bcn2 |
| 2 | SGDR T₀=40, T_mult=2 | 42 | 0.4063 | 13.3 | 8.3 | 32.8 | 6.9 | ggwj7vm9 |
| 3 | SGDR T₀=40, T_mult=2 | 43 | 0.4023 | 13.9 | 8.5 | 31.4 | 6.7 | cku6s4z2 |
| 4 | SGDR T₀=20, T_mult=2 | 42 | 0.4008 | 14.0 | 8.3 | 32.2 | 6.9 | 69kozelw |
| 5 | SGDR T₀=20, T_mult=2 | 43 | 0.4042 | 13.6 | 8.3 | 32.8 | 6.8 | 1gxlmtqd |
| 6 | SGDR T₀=60, T_mult=2 | 42 | 0.3961 | 13.6 | 8.3 | 31.6 | 6.8 | te3twxrp |
| 7 | SGDR T₀=60, T_mult=2 | 43 | 0.3941 | 13.5 | 8.2 | 30.5 | 6.5 | ulqizbvh |

### Mean ± Std by Config

| Config | val/loss | p_in | p_oodc | p_tan | p_re |
|--------|----------|------|--------|-------|------|
| **Baseline** | **0.3891 ± 0.001** | **13.00 ± 0.42** | **7.85 ± 0.35** | **30.70 ± 0.42** | **6.50 ± 0.14** |
| SGDR T₀=20 | 0.4025 ± 0.002 | 13.80 ± 0.28 | 8.30 ± 0.00 | 32.50 ± 0.42 | 6.85 ± 0.07 |
| SGDR T₀=40 | 0.4043 ± 0.003 | 13.60 ± 0.42 | 8.40 ± 0.14 | 32.10 ± 0.99 | 6.80 ± 0.14 |
| SGDR T₀=60 | 0.3951 ± 0.001 | 13.55 ± 0.07 | 8.25 ± 0.07 | 31.05 ± 0.78 | 6.65 ± 0.21 |

### Comparison vs Single-Model Baseline (8-seed mean from BASELINE.md)

| Metric | Baseline (8-seed mean) | Best SGDR (T₀=60) | Δ |
|--------|----------------------|-------------------|---|
| p_in | 13.03 | 13.55 | **+4.0%** |
| p_oodc | 7.83 | 8.25 | **+5.4%** |
| p_tan | 30.29 | 31.05 | **+2.5%** |
| p_re | 6.45 | 6.65 | **+3.1%** |

### Key Observations

**All SGDR variants underperform the baseline.** Even the best SGDR setting (T₀=60) is worse across every metric. The ranking among SGDR settings:

1. **T₀=60 (best SGDR):** Closest to baseline. With T_mult=2, the first cycle spans epochs 10-69, second starts at epoch 70. This most closely resembles the sequential cosine (single long descent), which explains why it's the least damaged by restarts.

2. **T₀=40:** Cycle 1 ends at epoch 49, cycle 2 spans epochs 50-129, cycle 3 starts at epoch 130. The restart at epoch 50 disrupts training mid-convergence.

3. **T₀=20 (worst SGDR):** Most frequent restarts. Cycle boundaries at epochs 10, 30, 70, 150. The frequent LR spikes prevent the model from settling into good basins.

### Why SGDR Hurts

The hypothesis was that warm restarts would help the model escape local minima and find basins that generalize better to OOD splits (especially p_tan). Instead:

- **p_tan got worse (+2.5% to +5.9%)** — restarts don't improve OOD generalization for this architecture
- **p_oodc consistently degraded (+5-8%)** — the OOD-condition split benefits from smooth convergence
- **More restarts = worse results** — the T₀=20 > T₀=40 > T₀=60 ordering suggests the LR spikes actively harm the learned representations

The sequential cosine scheduler with T_max=160 already provides a good balance: a single smooth descent that allows the model to fully converge. The EMA (starting at epoch 140) then stabilizes the final model. With warm restarts, the LR spikes at restart boundaries disrupt both the model weights and the EMA averaging window.

### Best Epoch Analysis

- Baseline: best at epoch 157 (near end of training)
- SGDR T₀=40: best at epoch **130** (second cycle boundary — model picks the restart point as best, suggesting it never improves after)
- SGDR T₀=20: best at epoch **151** 
- SGDR T₀=60: best at epoch **157**

The T₀=40 models selecting epoch 130 (start of cycle 3) as best confirms that post-restart performance degrades — the model was better before the restart.

### Commands Used

```bash
# Baseline (GPUs 0-1)
CUDA_VISIBLE_DEVICES={0,1} python train.py --agent askeladd --wandb_group phase6/sgdr-warm-restarts \
  --wandb_name "askeladd/baseline-s{42,43}" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed {42,43}

# SGDR (GPUs 2-7)
CUDA_VISIBLE_DEVICES={2-7} python train.py --agent askeladd --wandb_group phase6/sgdr-warm-restarts \
  --wandb_name "askeladd/sgdr-T0{20,40,60}-s{42,43}" \
  --scheduler_type warm_restarts --cosine_T_0 {20,40,60} --cosine_T_mult 2 \
  [same baseline flags as above, WITHOUT --cosine_T_max 160] \
  --seed {42,43}
```

W&B group: `phase6/sgdr-warm-restarts`

### Suggested Follow-ups

- **SGDR is a dead end for this architecture.** The sequential cosine scheduler is clearly superior. No further warm restart experiments are recommended.
- If more LR schedule exploration is desired, consider: (a) longer `cosine_T_max` (200, 250) with more epochs, or (b) cyclical cosine with very small restart amplitude (e.g. only bump LR to 50% of original, not 100%).
- The OOD generalization gap (p_tan being 2-3x worse than p_in) appears to be a fundamental architecture/data issue, not a training schedule issue.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2095 because ADVISOR: Closing as dead end. All 8 SGDR warm restart runs (T_0=20/40/60, seeds 42/73) completed — every variant produced worse p_oodc than baseline (8.19-8.47 vs 7.59-8.13 baseline). SGDR restarts hurt OOD generalization in this setup, likely because the restart disrupts the stable descent into the flat basin that Lion+cosine finds naturally. Adding to confirmed dead ends..
